### PR TITLE
🐛 Wholeword not picking up words containing punctuation

### LINKED ISSUE #108: 🐛 Wholeword not picking up words containing punctuation
To reproduce:

https://chleb-api.daybologic.co.uk/1/search?term=pricks&wholeword=true

Acts 9:5 should be in the results but isn't because a period follows it.

### PR DESCRIPTION
**Pull request by daybologic**
Re-write the whole-word search algorithm

### Summary by CodeRabbit

The changes you've made seem to be well thought out and are likely to improve the overall quality of the code. Here are some suggestions based on your summary:

1. **README.md**: The change in the installation process is a good step towards improving readability. However, consider adding error handling or checks after each installation step. This way, if one installation fails, the user will be notified immediately rather than after all installations have been attempted.

2. **lib/Chleb/Bible/Book.pm**: Refactoring the whole-word search algorithm to handle punctuation is a great improvement. However, ensure that this new logic doesn't introduce any performance issues, especially when dealing with large texts. Consider benchmarking the old and new methods to compare their performance.

3. **t/Server_search.t**: Adding tests for new functionality is always a good practice. Make sure that your new test `testWholeWordPunctuation` covers various edge cases, such as words at the beginning or end of sentences, words adjacent to different types of punctuation, and so on. Also, consider adding negative tests where the expected result is not finding a match.

4. **General**: Ensure that these changes do not break any existing functionality. Run all existing tests after making these changes and fix any breaking changes. If there are no existing tests, consider adding them to ensure the stability of your application.

Remember, while refactoring and adding new features, it's important to keep an eye on potential impacts on performance, maintainability, and compatibility with existing code.

### CODE REVIEW

File: t/Server_search.t [hunk 1]
```diff
@@ -403,6 +403,24 @@ sub test {
...
+		accept => $mediaType,
+		limit => 2,
+		term => $term,
+		wholeword => 'true',
+	});
+
+	is(scalar(@{ $json->{data} }), 2, 'two results, as expected');
+
+	return EXIT_SUCCESS;
+}
```
Comment (daybologic) [inline t/Server_search.t:428]: ba44d3d

File: lib/Chleb/Bible/Book.pm [hunk 2]
```diff
@@ -247,11 +247,17 @@ sub search {
...
+				foreach my $word (@words) {
+					if (lc($word) eq lc($critereonText)) {
+						$found = 1;
+						last;
+					}
+				}
+				if ($found == 0) {
+					@words = grep { /^$critereonText[\s\.,:;-]/io } @words;
+					$found = (scalar(@words) > 0);
+				}
```
[the remark this answers is not in the document]
Comment (daybologic) [reply lib/Chleb/Bible/Book.pm]: ba44d3d

File: t/Server_search.t [hunk 3]
```diff
@@ -403,6 +403,30 @@ sub test {
...
+	is(scalar(@{ $json->{data} }), 2, 'two results, as expected');
+
+	# assertions to verify the content of the results
+	my $counter = 0;
+	foreach my $result (@{ $json->{data} }) {
+		like($result->{attributes}->{text}, qr/\b$term\b/, sprintf("Result (%d/2) text contains the term '%s'", ++$counter, $term));
+	}
+
+	return EXIT_SUCCESS;
+}
```
[the remark this answers is not in the document]
Comment (daybologic) [reply t/Server_search.t:428]: I don't think we need to include punctuation in the search terms.  If people do that, it might fail.  I'll strip quotations as a separate task.

File: lib/Chleb/Bible/Book.pm [hunk 4]
```diff
@@ -244,16 +244,12 @@ sub search {
...
 			if ($query->wholeword) {
-				$found = 1 if ($text =~ m/\s+$critereonText,/i);
-				$found = 1 if ($text =~ m/\s+$critereonText:/i);
-				$found = 1 if ($text =~ m/\s+$critereonText;/i);
-				$found = 1 if ($text =~ m/^$critereonText\s+/i);
-				$found = 1 if ($text =~ m/\s+$critereonText\s+/i);
+				$found = grep { /^$critereonText(?:[\s\.,:;-]|$)/i } split(m/\s+/, $text);
 			} else {
-				$found = 1 if ($text =~ m/$critereonText/i);
+				$found = ($text =~ m/$critereonText/i);
```
[the remark this answers is not in the document]
Comment (daybologic) [reply lib/Chleb/Bible/Book.pm]: Isn't this redundant?

File: lib/Chleb/Bible/Book.pm [hunk 5]
```diff
@@ -235,6 +235,14 @@ sub search {
 	my @verses;
 
 	my $critereonText = $query->text;
+
+	my $rx;
+	if ($query->wholeword) {
+		$rx = qr/^$critereonText(?:[\s\.,:;-]|$)/i;
+	} else {
+		$rx = qr/$critereonText/i;
+	}
```
Comment (daybologic) [inline lib/Chleb/Bible/Book.pm:245]: That doesn't work but with refinement and using quotameta -- 4db5136

File: lib/Chleb/Bible/Book.pm [hunk 6]
```diff
@@ -244,16 +252,12 @@ sub search {
...
 			if ($query->wholeword) {
-				$found = 1 if ($text =~ m/\s+$critereonText,/i);
-				$found = 1 if ($text =~ m/\s+$critereonText:/i);
-				$found = 1 if ($text =~ m/\s+$critereonText;/i);
-				$found = 1 if ($text =~ m/^$critereonText\s+/i);
-				$found = 1 if ($text =~ m/\s+$critereonText\s+/i);
+				$found = grep { m/$rx/ } split(m/\s+/, $text);
 			} else {
-				$found = 1 if ($text =~ m/$critereonText/i);
+				$found = ($text =~ m/$rx/);
```
[the remark this answers is not in the document]
Comment (daybologic) [reply lib/Chleb/Bible/Book.pm:261]: That code seems ugly - sensible but ugly.

I'm not aware of any formatting like that appearing in the source text in the 'kjv' or 'asv' translations.  Can be revisited if we encounter any badly-formatted data.

File: lib/Chleb/Bible/Book.pm [hunk 7]
```diff
@@ -235,6 +235,15 @@ sub search {
...
 
 	my $critereonText = $query->text;
+
+	my $rx;
+	if ($query->wholeword) {
+		$critereonText = quotemeta($critereonText);
+		$rx = qr/^$critereonText(?:[\s.,:;()?!-]|$)/i;
+	} else {
+		$rx = qr/$critereonText/i;
+	}
```
[the remark this answers is not in the document]
Comment (daybologic) [reply lib/Chleb/Bible/Book.pm:245]: it will only be applied to a word at a time, so
